### PR TITLE
Change coords compression algorithm.

### PR DESCRIPTION
**Pull request by siarheishut**
Implement new algorithm to compress coords.

### CODE REVIEW

File: src/map_renderer.cpp [hunk 1]
```diff
@@ -84,16 +84,18 @@ std::unordered_map<std::string_view, svg::Point> TransformCoords(
   for (auto &bus : buses) {
     stops = Interpolate(std::move(stops), bus.second.route, base_stops);
   }
+
   auto adjacent_stops = AdjacentStops(buses);
+  auto adjacent_list = AdjacentList(adjacent_stops);
 
   auto stops_by_lon = SortStops(stops, SortMode::kByLongitude);
-  auto layers_by_lon = CompressNonadjacent(stops_by_lon, adjacent_stops);
+  auto layers_by_lon = CompressNonNeighbors(stops_by_lon, adjacent_list);
```
Comment (foxfrikses) [inline src/map_renderer.cpp:92]: Okay, I've told you that you can just create `AdjacentList` function with intention to use it inside `CompressNonadjacent` in order to not change the interface of the function.

But I can see now, that you need this adjacent list multiple times, so now I would say that it'd be better to make `AdjacentStops` return the adjacent list instead of a list of the neighbors.
I wold also recommend creating an alias for the adjacent list type, like `AdjacentList` to make it more readable.

File: src/coords_converter.h [hunk 2]
```diff
@@ -47,10 +47,15 @@ renderer_utils::Stops Interpolate(
 std::vector<std::pair<std::string_view, std::string_view>>
 AdjacentStops(const renderer_utils::Buses &buses);
 
-std::vector<std::vector<std::string_view>> CompressNonadjacent(
+std::unordered_map<std::string_view, std::vector<std::string_view>>
+AdjacentList(const std::vector<std::pair<std::string_view,
+                                         std::string_view>> &neighbors);
+
+std::vector<std::vector<std::string_view>>
+CompressNonNeighbors(
```
Comment (foxfrikses) [inline src/coords_converter.h:55]: I don't think that you need to rename this function, do you?

File: src/coords_converter.cpp [hunk 3]
```diff
@@ -119,35 +119,47 @@ AdjacentStops(const renderer_utils::Buses &buses) {
...
-  vector<int> max_next(stops.size(), static_cast<int>(stops.size()) - 1);
+  if (stop_to_layer.empty()) return {};
 
-  for (auto &[prev, curr] : neighbors) {
-    auto prev_layer = stop_to_layer[prev];
-    auto curr_layer = stop_to_layer[curr];
-    if (prev_layer > curr_layer) swap(prev_layer, curr_layer);
-    max_next[prev_layer] = min(max_next[prev_layer], curr_layer - 1);
-  }
+  auto layers_count = max_element(begin(stop_to_layer), end(stop_to_layer),
```
Comment (foxfrikses) [inline src/coords_converter.cpp:153]: `layer_count`

File: src/coords_converter.cpp [hunk 4]
```diff
@@ -119,35 +119,47 @@ AdjacentStops(const renderer_utils::Buses &buses) {
...
+    auto it_neighbors = adj_list.find(stop);
+    if (it_neighbors != adj_list.end()) {
+      for (auto &neighbor : it_neighbors->second) {
+        if (auto found = stop_to_layer.find(neighbor);
+            found != stop_to_layer.end())
+          id = max(id, found->second + 1);
+      }
+    }
+    stop_to_layer[stop] = id;
+  }
```
Comment (foxfrikses) [inline src/coords_converter.cpp:149]: You can reduce the number or nested blocks here: 
```
for (auto &stop : stops) {
  auto& layer = stop_to_layer[stop] = 0;
  auto it = adj_list.find(stop);
  if (it == adj_list.end()) continue;
  for (auto &neighbor : it->second) {
    auto found = stop_to_layer.find(neighbor);
    if (found == stop_to_layer.end()) continue;
    layer = max(layer, found->second + 1);
  }
}
```

File: src/coords_converter.h [hunk 5]
```diff
@@ -47,10 +47,15 @@ renderer_utils::Stops Interpolate(
...
     const std::vector<std::string_view> &stops,
-    const std::vector<std::pair<std::string_view,
-                                std::string_view>> &neighbors);
+    const std::unordered_map<std::string_view,
+                             std::vector<std::string_view>> &adj_list);
```
Comment (foxfrikses) [inline src/coords_converter.h:59]: Please use an alias here like `AdjacentList`. And I would name this arg as `adj_stops`

File: src/map_renderer.cpp (hunk 1 above)
Comment (siarheishut) [reply src/map_renderer.cpp:92]: Done.

File: src/coords_converter.cpp (hunk 3 above)
Comment (siarheishut) [reply src/coords_converter.cpp:153]: Done.

File: src/coords_converter.h (hunk 2 above)
Comment (siarheishut) [reply src/coords_converter.h:55]: Done.

File: src/coords_converter.cpp (hunk 4 above)
Comment (siarheishut) [reply src/coords_converter.cpp:149]: Done.

File: src/coords_converter.h (hunk 5 above)
Comment (siarheishut) [reply src/coords_converter.h:59]: Done.

File: src/coords_converter.cpp [hunk 6]
```diff
@@ -105,49 +105,45 @@ renderer_utils::Stops Interpolate(
...
-    for (int i = 1; i < route.route.size(); ++i) {
-      auto prev = route.route[i - 1];
-      auto curr = route.route[i];
-      if (prev == curr) continue;
-      result.emplace_back(prev, curr);
+AdjacentList AdjacentStops(const renderer_utils::Buses &buses) {
+  AdjacentList adj_stops;
+  for (auto &bus : buses) {
+    auto &route = bus.second.route;
+    for (int i = 0; i < route.size(); ++i) {
```
Comment (foxfrikses) [inline src/coords_converter.cpp:112]: Why not just iterate from 1 instead of 0 and get rid of all the if-s

File: src/map_renderer.cpp [hunk 7]
```diff
@@ -84,16 +84,17 @@ std::unordered_map<std::string_view, svg::Point> TransformCoords(
...
   auto stops_by_lon = SortStops(stops, SortMode::kByLongitude);
-  auto layers_by_lon = CompressNonadjacent(stops_by_lon, adjacent_stops);
+  auto layers_by_lon = CompressNonadjacent(stops_by_lon, adj_stops);
   auto x_coords = SpreadStops(layers_by_lon,
                               frame.padding,
                               frame.width - frame.padding);
 
   auto stops_by_lat = SortStops(stops, SortMode::kByLatitude);
-  auto layers_by_lat = CompressNonadjacent(stops_by_lat, adjacent_stops);
+  auto layers_by_lat = CompressNonadjacent(stops_by_lat, adj_stops);
```
Comment (foxfrikses) [inline src/map_renderer.cpp:97]: I think no modifications of this file are required.

File: src/coords_converter.cpp (hunk 6 above)
Comment (siarheishut) [reply src/coords_converter.cpp:112]: Done.

File: src/map_renderer.cpp (hunk 7 above)
Comment (siarheishut) [reply src/map_renderer.cpp:97]: Fixed.

File: src/map_renderer.cpp [hunk 8]
```diff
@@ -84,6 +84,7 @@ std::unordered_map<std::string_view, svg::Point> TransformCoords(
   for (auto &bus : buses) {
     stops = Interpolate(std::move(stops), bus.second.route, base_stops);
   }
+
```
Comment (foxfrikses) [inline src/map_renderer.cpp:87]: No need for this change in this commit

File: tests/coords_converter_test.cpp [hunk 9]
```diff
@@ -86,68 +86,86 @@ TEST(TestAdjacentStops, TestAdjacentStops) {
...
   for (auto &[name, buses, want] : test_cases) {
     auto got = rm::coords_converter::AdjacentStops(buses);
+    for (auto &[_, neighbors] : got) {
+      sort(begin(neighbors), end(neighbors));
+    }
+    for (auto &[_, neighbors] : want) {
+      sort(begin(neighbors), end(neighbors));
+    }
+
     EXPECT_EQ(want, got) << name;
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:120]: Why not [UnorderedElementsAre](http://google.github.io/googletest/reference/matchers.html#container-matchers)?

Comment (foxfrikses) [reply tests/coords_converter_test.cpp:120]: Okay, never mind, it won't work here, but what will work is `operator ==` and `operator <` for `rm::coords_converter::AdjacentList`, what do you think?

File: src/coords_converter.h [hunk 10]
```diff
@@ -44,13 +47,11 @@ renderer_utils::Stops Interpolate(
...
 
-std::vector<std::pair<std::string_view, std::string_view>>
-AdjacentStops(const renderer_utils::Buses &buses);
+AdjacentList AdjacentStops(const renderer_utils::Buses &buses);
 
-std::vector<std::vector<std::string_view>> CompressNonadjacent(
-    const std::vector<std::string_view> &stops,
-    const std::vector<std::pair<std::string_view,
-                                std::string_view>> &neighbors);
+std::vector<std::vector<std::string_view>>
```
Comment (foxfrikses) [inline src/coords_converter.h:52]: I think you can create an alias for this as well, like `StopLayers`, what do you think about it?

File: tests/coords_converter_test.cpp [hunk 11]
```diff
@@ -86,68 +86,86 @@ TEST(TestAdjacentStops, TestAdjacentStops) {
...
     rm::renderer_utils::Buses buses;
-    vector<pair<string_view, string_view>> want;
+    rm::coords_converter::AdjacentList want;
   };
 
   vector<TestCase> test_cases{
       TestCase{
-          .name = "Empty buses",
-          .want = {},
-      },
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:96]: Why did you delete `Empty buses` test case?

File: tests/coords_converter_test.cpp [hunk 12]
```diff
@@ -86,68 +86,86 @@ TEST(TestAdjacentStops, TestAdjacentStops) {
...
+          .want = {{"a", {"b", "d"}}, {"b", {"a", "c"}}, {"c", {"b", "d"}},
+                   {"d", {"a", "c"}}}
       },
       TestCase{
-          .name = "Repeated stop",
-          .buses = {{"1", {{"1", "2", "2", "3"}, true}}},
-          .want = {{"1", "2"}, {"2", "3"}},
-      },
+          .name = "Several buses",
+          .buses = {{"bus 1", {{"a", "b", "c", "d", "a"}, true}},
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:101]: What if there are two identical stops following one another, like route `"a", "b", "b", "c", "d"`? I think, you'll have a wrong logic here.

File: tests/coords_converter_test.cpp [hunk 13]
```diff
@@ -86,68 +86,86 @@ TEST(TestAdjacentStops, TestAdjacentStops) {
...
+                    {"bus 2", {{"e", "f", "b", "g", "h"}, false}},
+                    {"bus 3", {{"a", "i", "g", "j", "b"}, false}}},
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:103]: What if `"a"` and `"b"` are neighbouring twice in one route? Or if they are neighbouring in multiple routes? Now you'll have them being present twice in every occurrence.

Comment (foxfrikses) [reply tests/coords_converter_test.cpp:103]: Is  `vector` and `n * push_back` + `sort` + `unique` + `erase` better than `unordered_map` and `n * insert`?

File: src/coords_converter.h [hunk 14]
```diff
@@ -44,13 +47,11 @@ renderer_utils::Stops Interpolate(
...
+CompressNonadjacent(
+    const std::vector<std::string_view> &stops, const AdjacentList &adj_stops);
```
Comment (foxfrikses) [inline src/coords_converter.h:54]: I think you should add a comment that this function expects all the stops to be unique, otherwise the behaviour is not defined.

Comment (siarheishut) [reply src/coords_converter.h:54]: Done.

File: tests/coords_converter_test.cpp (hunk 13 above)
Comment (siarheishut) [reply tests/coords_converter_test.cpp:103]: std::unordered_set has better performance then std::vector here.

File: tests/coords_converter_test.cpp (hunk 12 above)
Comment (siarheishut) [reply tests/coords_converter_test.cpp:101]: Fixed.

File: tests/coords_converter_test.cpp (hunk 11 above)
Comment (siarheishut) [reply tests/coords_converter_test.cpp:96]: Added.

File: src/coords_converter.h (hunk 10 above)
Comment (siarheishut) [reply src/coords_converter.h:52]: Done.

File: tests/coords_converter_test.cpp (hunk 9 above)
Comment (siarheishut) [reply tests/coords_converter_test.cpp:120]: Fixed.

File: src/map_renderer.cpp (hunk 8 above)
Comment (siarheishut) [reply src/map_renderer.cpp:87]: Removed.

File: tests/coords_converter_test.cpp (hunk 12 above)
Comment (foxfrikses) [reply tests/coords_converter_test.cpp:108]: What about a test?

Comment (siarheishut) [reply tests/coords_converter_test.cpp:113]: Added.

File: tests/coords_converter_test.cpp [hunk 15]
```diff
@@ -95,60 +98,87 @@ TEST(TestAdjacentStops, TestAdjacentStops) {
...
+                   {"f", {"b", "e"}}, {"g", {"b", "h", "i", "j"}}, {"h", {"g"}},
+                   {"i", {"a", "g"}}, {"j", {"b", "g"}}},
+      }
   };
 
   for (auto &[name, buses, want] : test_cases) {
+    using namespace ::testing;
+
     auto got = rm::coords_converter::AdjacentStops(buses);
+
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:127]: Miss

File: tests/test_utils.h [hunk 16]
```diff
@@ -6,8 +6,10 @@
 #include <vector>
 
 #include "svg/common.h"
+#include "gmock/gmock.h"
```
Comment (foxfrikses) [inline tests/test_utils.h:9]: No change in this file is required

File: tests/test_utils.cpp [hunk 17]
```diff
@@ -1,13 +1,14 @@
...
 #include <iomanip>
 #include <sstream>
 #include <string>
+#include <string_view>
 #include <tuple>
 #include <vector>
 
-#include "svg/common.h"
-
+#include "src/coords_converter.h"
```
Comment (foxfrikses) [inline tests/test_utils.cpp:11]: No change in this file is required

File: tests/test_utils.h [hunk 18]
```diff
@@ -8,6 +8,7 @@
 #include "svg/common.h"
 
 #include "src/bus_manager.h"
+#include "src/coords_converter.h"
```
Comment (foxfrikses) [inline tests/test_utils.h:11]: -

**Review by foxfrikses — APPROVED** [review]
LGTM 💯
